src/vm/loaderallocator.cpp

```diff
@@ -2046,4 +2053,126 @@ BOOL LoaderAllocator::InsertComInteropData(MethodTable* pMT, InteropMethodTableD
 
 #endif // FEATURE_COMINTEROP
 
+
+void LoaderAllocator::AddUnmanagedImageToCache(LPCWSTR libraryName, HMODULE hMod)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        PRECONDITION(CheckPointer(libraryName));
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACTL_END;
+    if (libraryName)
+    {
+        m_UnmanagedCache.InsertEntry(libraryName, hMod);
+    }
+    return ;
+}
+
+HMODULE LoaderAllocator::FindUnmanagedImageInCache(LPCWSTR libraryName)
+{
+    CONTRACT(HMODULE)
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        PRECONDITION(CheckPointer(libraryName,NULL_OK));
+        POSTCONDITION(CheckPointer(RETVAL,NULL_OK));
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACT_END;
+    if(libraryName == NULL) RETURN NULL;
+
+    RETURN m_UnmanagedCache.LookupEntry(libraryName);
+}
+
+/* static */
+BOOL UnmanagedImageCache::CompareBindingSpec(UPTR spec1, UPTR spec2)
+{
+    WRAPPER_NO_CONTRACT;
+
+    LPCWSTR libraryName = (LPCWSTR) (spec1 << 1);
+    ImageEntry* pEntry2 = (ImageEntry*) spec2;
+
+    return wcscmp(libraryName, pEntry2->libraryName) == 0;
+}
+
+HMODULE UnmanagedImageCache::LookupEntry(LPCWSTR libraryName)
+{
+    CONTRACT (HMODULE)
+    {
+        INSTANCE_CHECK;
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        POSTCONDITION(CheckPointer(RETVAL, NULL_OK));
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACT_END
+
+    DWORD hashValue = HashString(libraryName);
+
+    LPVOID pResult = m_Table.LookupValue(hashValue, (LPVOID)libraryName);
+    if(pResult == (LPVOID) INVALIDENTRY)
+        RETURN NULL;
+    else
+        RETURN ((ImageEntry*)pResult)->hMod;
+}
+
+void UnmanagedImageCache::InsertEntry(LPCWSTR libraryName, HMODULE hMod)
+{
+    CONTRACTL
+    {
+        INSTANCE_CHECK;
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACTL_END
+
+    HMODULE hCachedMod = LookupEntry(libraryName);
+    if(hCachedMod == NULL)
+    {
+        LoaderAllocator::CacheLockHolder lh(m_pLoaderAllocator);
+
+        hCachedMod = LookupEntry(libraryName);
+        if (hCachedMod == NULL) 
+        {
+            AllocMemTracker amTracker;
+            AllocMemTracker *pamTracker = &amTracker;
+
+            ImageEntry* pEntry = (ImageEntry*) pamTracker->Track( m_pLoaderAllocator->GetLowFrequencyHeap()->AllocMem(S_SIZE_T(sizeof(ImageEntry))) );
+
+            size_t len = wcslen(libraryName) + 1;
+            LPWSTR temp = (LPWSTR)pamTracker->Track( m_pLoaderAllocator->GetLowFrequencyHeap()->AllocMem(S_SIZE_T (len * 2)) );
+            wcscpy_s(temp, len, libraryName);
+            pEntry->libraryName = temp;
+            pEntry->hMod = hMod;
+            DWORD hashValue = HashString(libraryName);
+            m_Table.InsertValue(hashValue, pEntry);
+
+            pamTracker->SuppressRelease();
+        }
+        // lh goes out of scope here
+    }
+#ifdef _DEBUG
```

Comment:
This `#ifdef` condition can be removed. The compiler will optimize out the empty branches on release builds, or assert that `hCachedMod == NULL`?